### PR TITLE
Reuse sending port of Interconnect across different queries

### PR DESCRIPTION
**Pull request by eurekaka**
Azure has restriction on distinct <src_port, dst_port> pairs: 500k per 5 minutes. Currently, GPDB backend process reuses IC listener port for its lifetime, while setting up a new IC sending port for each query, which makes it easy to hit this limit for commands like `analyzedb`.

### CODE REVIEW

**Comment by paul-guo-** [comment]
Does this introduce any performance regression?  I have the concern because under high concurrency there could be kernel contention for same socket pair. Search the context of SO_REUSEPORT for more related information.

**Comment by eurekaka** [comment]
@paul-guo- No performance regression I think, previously, for each query, we create socket and close it in TeardownInterconnect, while now we keep this socket for the lifetime of the process. SO_REUSEPORT is for port re-usage after the socket is closed IIRC, so I think it's not relevant to this issue. We are not closing the socket and then create a new socket to bind to the previous port, but keeping the socket across queries.

**Comment by paul-guo-** [comment]
SO_REUSEPORT is not for port reuse after closing. Here is a simple introduction: https://lwn.net/Articles/542629/. You could just check the UDP part. It is used to resolve the kernel socket related lock contention.

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -562,6 +562,10 @@ struct UnackQueueRing
  */
 static UnackQueueRing unack_queue_ring = {0, 0, 0};
 
+static int ICSenderSocket = -1;
```
Comment (hlinnaka) [inline src/backend/cdb/motion/ic_udpifc.c]: A comment to explain what this socket is for, would be nice.

**Comment by hlinnaka** [comment]
@eurekaka, I think you were thinking of SO_REUSEADDR. That's the one to allow quick reuse of a port, after the socket is closed.

SO_REUSEPORT is different. It allows to sockets to be bound to the same port, and both to listen actively. The kernel will then distribute incoming packets to both sockets, more or less evenly. That allows you to scale the listener, by having multiple threads or processes receive incoming packets on the same port. But I don't think that's relevant to the interconnect at all.

I'm not worried about performance regressions with this. With or without this patch, a backend process has one sender socket open at a time. The only difference is that we don't close and reopen repeatedly.

**Comment by eurekaka** [comment]
Thanks for the clarification, merged

**Pull request merged**